{
    "assignee": null,
    "assignees": [],
    "body": "Given the following extension class\r\n```cython\r\ncdef class PolynomialParameters(object):\r\n    cdef:\r\n        public list elementary_symmetric_polynomial\r\n        public list power_sum\r\n    def __init__(self, elementary_symmetric_polynomial, power_sum):\r\n        self.elementary_symmetric_polynomial = elementary_symmetric_polynomial\r\n        self.power_sum = power_sum\r\n\r\n    def __iter__(self):\r\n        yield self.power_sum\r\n        yield self.elementary_symmetric_polynomial\r\n```\r\nCython 0.26 generates the following code for the class's `__iter__` method:\r\n```c\r\nstatic PyObject *__pyx_gb_7brainpy_8_speedup_20PolynomialParameters_4generator(__pyx_CoroutineObject *__pyx_generator, PyObject *__pyx_sent_value) /* generator body */\r\n{\r\n  struct __pyx_obj_7brainpy_8_speedup___pyx_scope_struct____iter__ *__pyx_cur_scope = ((struct __pyx_obj_7brainpy_8_speedup___pyx_scope_struct____iter__ *)__pyx_generator->closure);\r\n  PyObject *__pyx_r = NULL;\r\n  __Pyx_RefNannyDeclarations\r\n  __Pyx_RefNannySetupContext(\"None\", 0);\r\n  __Pyx_TraceDeclarations\r\n  __Pyx_TraceCall(\"__iter__\", __pyx_f[0], 105, 0, __PYX_ERR(0, 105, __pyx_L1_error));\r\n  switch (__pyx_generator->resume_label) {\r\n    case 0: goto __pyx_L3_first_run;\r\n    case 1: goto __pyx_L4_resume_from_yield;\r\n    case 2: goto __pyx_L5_resume_from_yield;\r\n    default: /* CPython raises the right error here */\r\n    __Pyx_TraceReturn(Py_None, 0);\r\n    __Pyx_RefNannyFinishContext();\r\n    return NULL;\r\n  }\r\n //gotos....\r\n```\r\nThe line `__Pyx_TraceDeclarations` looks like it was meant to be a macro, given that it has no trailing semi-colon, but MSVC appears to be seeing it literally and throws the error `error C2143: syntax error : missing ';' before 'type'`. \r\n\r\nCython 0.25.2 generated the following code for the same method\r\n```c\r\nstatic PyObject *__pyx_gb_7brainpy_8_speedup_20PolynomialParameters_4generator(__pyx_CoroutineObject *__pyx_generator, PyObject *__pyx_sent_value) /* generator body */\r\n{\r\n  struct __pyx_obj_7brainpy_8_speedup___pyx_scope_struct____iter__ *__pyx_cur_scope = ((struct __pyx_obj_7brainpy_8_speedup___pyx_scope_struct____iter__ *)__pyx_generator->closure);\r\n  PyObject *__pyx_r = NULL;\r\n  __Pyx_RefNannyDeclarations\r\n  __Pyx_RefNannySetupContext(\"None\", 0);\r\n  switch (__pyx_generator->resume_label) {\r\n    case 0: goto __pyx_L3_first_run;\r\n    case 1: goto __pyx_L4_resume_from_yield;\r\n    case 2: goto __pyx_L5_resume_from_yield;\r\n    default: /* CPython raises the right error here */\r\n    __Pyx_RefNannyFinishContext();\r\n    return NULL;\r\n  }\r\n //gotos...\r\n```\r\nwhich compiles successfully with the same setup.\r\n\r\nThere is no issue on Linux. Could it be that `__Pyx_TraceDeclarations` is only defined inside a Unix-specific preprocessor block?\r\n\r\nThe repository this code is from is https://github.com/mobiusklein/brainpy, and the file being compiled is `brainpy/_speedup.pyx` in the course of the standard `python setup.py install` process",
    "closed_at": "2017-07-31T08:57:45Z",
    "comment_data": [
        {
            "body": "Thank you for the very detailed report.",
            "created_at": "2017-07-31T08:58:54Z",
            "html_url": "https://github.com/cython/cython/issues/1801#issuecomment-319009101",
            "id": 319009101,
            "issue_url": "https://api.github.com/repos/cython/cython/issues/1801",
            "updated_at": "2017-07-31T08:58:54Z",
            "url": "https://api.github.com/repos/cython/cython/issues/comments/319009101",
            "user": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            }
        }
    ],
    "comments": 1,
    "comments_url": "https://api.github.com/repos/cython/cython/issues/1801/comments",
    "created_at": "2017-07-28T15:19:39Z",
    "event_data": [
        {
            "actor": {
                "avatar_url": "https://avatars0.githubusercontent.com/u/491659?v=4",
                "events_url": "https://api.github.com/users/scoder/events{/privacy}",
                "followers_url": "https://api.github.com/users/scoder/followers",
                "following_url": "https://api.github.com/users/scoder/following{/other_user}",
                "gists_url": "https://api.github.com/users/scoder/gists{/gist_id}",
                "gravatar_id": "",
                "html_url": "https://github.com/scoder",
                "id": 491659,
                "login": "scoder",
                "organizations_url": "https://api.github.com/users/scoder/orgs",
                "received_events_url": "https://api.github.com/users/scoder/received_events",
                "repos_url": "https://api.github.com/users/scoder/repos",
                "site_admin": false,
                "starred_url": "https://api.github.com/users/scoder/starred{/owner}{/repo}",
                "subscriptions_url": "https://api.github.com/users/scoder/subscriptions",
                "type": "User",
                "url": "https://api.github.com/users/scoder"
            },
            "commit_id": "4b61de606e05fb9c9a7eb4cdf3e897a4740e1f27",
            "commit_url": "https://api.github.com/repos/cython/cython/commits/4b61de606e05fb9c9a7eb4cdf3e897a4740e1f27",
            "created_at": "2017-07-31T08:57:45Z",
            "event": "closed",
            "id": 1185363936,
            "url": "https://api.github.com/repos/cython/cython/issues/events/1185363936"
        }
    ],
    "events_url": "https://api.github.com/repos/cython/cython/issues/1801/events",
    "html_url": "https://github.com/cython/cython/issues/1801",
    "id": 246378917,
    "labels": [],
    "labels_url": "https://api.github.com/repos/cython/cython/issues/1801/labels{/name}",
    "locked": false,
    "milestone": null,
    "number": 1801,
    "repository_url": "https://api.github.com/repos/cython/cython",
    "state": "closed",
    "title": "Cython 0.26 generates invalid C in extension type __iter__  methods for MSVC and Python 2.7",
    "updated_at": "2017-07-31T08:58:54Z",
    "url": "https://api.github.com/repos/cython/cython/issues/1801",
    "user": {
        "avatar_url": "https://avatars2.githubusercontent.com/u/1426041?v=4",
        "events_url": "https://api.github.com/users/mobiusklein/events{/privacy}",
        "followers_url": "https://api.github.com/users/mobiusklein/followers",
        "following_url": "https://api.github.com/users/mobiusklein/following{/other_user}",
        "gists_url": "https://api.github.com/users/mobiusklein/gists{/gist_id}",
        "gravatar_id": "",
        "html_url": "https://github.com/mobiusklein",
        "id": 1426041,
        "login": "mobiusklein",
        "organizations_url": "https://api.github.com/users/mobiusklein/orgs",
        "received_events_url": "https://api.github.com/users/mobiusklein/received_events",
        "repos_url": "https://api.github.com/users/mobiusklein/repos",
        "site_admin": false,
        "starred_url": "https://api.github.com/users/mobiusklein/starred{/owner}{/repo}",
        "subscriptions_url": "https://api.github.com/users/mobiusklein/subscriptions",
        "type": "User",
        "url": "https://api.github.com/users/mobiusklein"
    }
}